### PR TITLE
Debounce cleanups

### PR DESCRIPTION
**Pull request by mairas**
I did a bit of reworking on the networking logic and removed the "reboot on wifi disconnect" from the Networking class and made it a consumer of SystemState in SensESPApp. For this, I required Debounce to only trigger the reset if the state had been stable for 3 minutes.

First, I noticed that due to the Debounce template implementation, it was impossibly to instantiate Debounce<SystemStatus>. I did a quick benchmark to verify the claim in Issue #287 that "if you put the entire implementation in the header file (esp. the configuration stuff) the compiler will generate a TON of redundant code." (The issue sounded a bit peculiar because basically all of STL is, well, templates.)

I have a random SensESP test program that takes some changing AnalogInput, applies hysteresis and rounding and then Debounces the boolean and integer values. Compiled with the original implementation, the RAM footprint was 51.0 KB and flash footprint 1015.4 KB. I then refactored the code into a standard headers-only template class. The RAM footprint remained at 51.0 KB while the flash footprint dropped to 1011.0 KB. So, that is the first major change in this PR.

The second change relates to the Debouncing logic. I am debouncing the system state which gets regular WifiDisconnected updates (the same value all over again). The original logic reset the timer for every update, even if they repeated the same value. That didn't quite make sense to me - it would be impossible to debounce any polled values with this logic. I shuffled the code a bit around but the only change in the end was that repeated values are considered stable instead of resetting the timer every time.

### CODE REVIEW

File: src/transforms/debounce.h [hunk 1]
```diff
@@ -42,26 +42,31 @@ class DebounceTemplate : public SymmetricTransform<T> {
...
-        this->value_sent_ = true;
+        this->debounced_value_ = input;
         this->emit(input);
       });
+      value_received_ = true;
     }
   }
 
  private:
+  bool require_change_;
```
Comment (ba58smith) [inline src/transforms/debounce.h:66]: It doesn't appear that `bool require_change_;` is used anywhere.

**Review by ba58smith — APPROVED** [review]
Looks good to me.

File: src/transforms/debounce.h (hunk 1 above)
Comment (mairas) [reply src/transforms/debounce.h:66]: Ah! An accidental leftover. Will remove.

Comment (mairas) [reply src/transforms/debounce.h]: Fixed.

**Comment by mairas** [comment]
Since this PR includes a change in the behavior logic, if @joelkoz has any doubt, it'd be great if you could test it before merging so that you're not left in the dark (pun absolutely intended!).

**Pull request merged**